### PR TITLE
Functions Interop

### PR DESCRIPTION
**Pull request by bstpierr**
Convert Firebase Functions to use the Auth Interop pod

### CODE REVIEW

**Comment by bstpierr** [comment]
I now have Travis passing all the tests on this PR. Please have a look.

File: Functions/Example/FirebaseFunctions.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -483,6 +492,11 @@
 				GCC_WARN_UNINITIALIZED_AUTOS = YES_AGGRESSIVE;
 				GCC_WARN_UNUSED_FUNCTION = YES;
 				GCC_WARN_UNUSED_VARIABLE = YES;
+				HEADER_SEARCH_PATHS = (
+					"$(inherited)",
+					"$(SRCROOT)/../FirebaseFunctions/**",
```
[the remark this answers is not in the document]
Comment (bstpierr) [reply Functions/Example/FirebaseFunctions.xcodeproj/project.pbxproj]: Yes, I reverted the search path back to the public Functions header, and moved the recursive path to the test targets only.

File: Functions/Example/IntegrationTests/FIRIntegrationTests.m [hunk 2]
```diff
@@ -31,7 +32,10 @@ @implementation FIRIntegrationTests
 - (void)setUp {
   [super setUp];
   id app = [[FUNFakeApp alloc] initWithProjectID:@"functions-integration-test"];
-  _functions = [FIRFunctions functionsForApp:app];
+  _functions = [[FIRFunctions alloc]
```
[the remark this answers is not in the document]
Comment (bstpierr) [reply Functions/Example/IntegrationTests/FIRIntegrationTests.m]: Done.

**Comment by ryanwilson** [comment]
Looks like there's a Travis issue: https://travis-ci.org/firebase/firebase-ios-sdk/jobs/460759039#L7046

**Comment by bstpierr** [comment]
Yeah, seems I removed too much of my previous PR. I'm not 100% which path included the container header, but I'll fiddle with it and look at where CocoaPods puts everything.

File: Functions/Example/FirebaseFunctions.xcodeproj/project.pbxproj [hunk 3]
```diff
@@ -538,7 +550,13 @@
 					"DEBUG=1",
 					"$(inherited)",
 				);
-				HEADER_SEARCH_PATHS = "$(SRCROOT)/../FirebaseFunctions";
+				HEADER_SEARCH_PATHS = (
+					"$(inherited)",
+					"\"${PODS_ROOT}/Headers/Public\"",
+					"\"${PODS_ROOT}/Headers/Public/FirebaseAnalyticsInterop\"",
```
Comment (paulb777) [inline Functions/Example/FirebaseFunctions.xcodeproj/project.pbxproj]: delete

File: Functions/Example/FirebaseFunctions.xcodeproj/project.pbxproj [hunk 4]
```diff
@@ -557,7 +575,13 @@
 					"$(DEVELOPER_FRAMEWORKS_DIR)",
 				);
 				GCC_PRECOMPILE_PREFIX_HEADER = NO;
-				HEADER_SEARCH_PATHS = "$(SRCROOT)/../FirebaseFunctions";
+				HEADER_SEARCH_PATHS = (
+					"$(inherited)",
+					"\"${PODS_ROOT}/Headers/Public\"",
+					"\"${PODS_ROOT}/Headers/Public/FirebaseAnalyticsInterop\"",
```
Comment (paulb777) [inline Functions/Example/FirebaseFunctions.xcodeproj/project.pbxproj]: delete

File: Functions/Example/FirebaseFunctions.xcodeproj/project.pbxproj [hunk 5]
```diff
@@ -580,7 +604,13 @@
 					"DEBUG=1",
 					"$(inherited)",
 				);
-				HEADER_SEARCH_PATHS = "$(SRCROOT)/../FirebaseFunctions";
+				HEADER_SEARCH_PATHS = (
+					"$(inherited)",
+					"\"${PODS_ROOT}/Headers/Public\"",
+					"\"${PODS_ROOT}/Headers/Public/FirebaseAnalyticsInterop\"",
```
Comment (paulb777) [inline Functions/Example/FirebaseFunctions.xcodeproj/project.pbxproj]: delete

File: Functions/Example/FirebaseFunctions.xcodeproj/project.pbxproj [hunk 6]
```diff
@@ -580,7 +604,13 @@
...
+					"\"${PODS_ROOT}/Headers/Public/FirebaseAuthInterop\"",
+					"\"$(SRCROOT)/../FirebaseFunctions\"/**",
```
Comment (paulb777) [inline Functions/Example/FirebaseFunctions.xcodeproj/project.pbxproj]: Thanks for addressing the Xcode issue. Recursive search paths are maybe ok for tests only, but would be nice to avoid  - they can hide issues and cause tricky bugs.

File: Functions/Example/FirebaseFunctions.xcodeproj/project.pbxproj [hunk 7]
```diff
@@ -599,7 +629,13 @@
 					"$(DEVELOPER_FRAMEWORKS_DIR)",
 				);
 				GCC_PRECOMPILE_PREFIX_HEADER = NO;
-				HEADER_SEARCH_PATHS = "$(SRCROOT)/../FirebaseFunctions";
+				HEADER_SEARCH_PATHS = (
+					"$(inherited)",
+					"\"${PODS_ROOT}/Headers/Public\"",
+					"\"${PODS_ROOT}/Headers/Public/FirebaseAnalyticsInterop\"",
```
Comment (paulb777) [inline Functions/Example/FirebaseFunctions.xcodeproj/project.pbxproj]: delete

File: Functions/Example/Podfile [hunk 8]
```diff
@@ -1,12 +1,15 @@
 use_frameworks!
 
+pod 'FirebaseFunctions', :path => '../../'
+
+pod 'FirebaseAnalyticsInterop', :path => '../../'
```
Comment (paulb777) [inline Functions/Example/Podfile]: Why is AnalyticsInterop here? Doesn't functions only use Auth?

File: Functions/Example/FirebaseFunctions.xcodeproj/project.pbxproj [hunk 9]
```diff
@@ -494,13 +500,8 @@
 			isa = XCBuildConfiguration;
 			buildSettings = {
 				ASSETCATALOG_COMPILER_APPICON_NAME = AppIcon;
-				GCC_PRECOMPILE_PREFIX_HEADER = NO;
```
Comment (paulb777) [inline Functions/Example/FirebaseFunctions.xcodeproj/project.pbxproj]: Why this change?

File: Functions/Example/FirebaseFunctions.xcodeproj/project.pbxproj [hunk 10]
```diff
@@ -494,13 +500,8 @@
...
 				INFOPLIST_FILE = "FirebaseFunctions/FirebaseFunctions-Info.plist";
 				MODULE_NAME = ExampleApp;
-				OTHER_LDFLAGS = (
-					"$(inherited)",
-					"-all_load",
```
Comment (paulb777) [inline Functions/Example/FirebaseFunctions.xcodeproj/project.pbxproj]: -all_load should stay to catch more errors earlier

File: Functions/Example/FirebaseFunctions.xcodeproj/project.pbxproj (hunk 9 above)
Comment (bstpierr) [reply Functions/Example/FirebaseFunctions.xcodeproj/project.pbxproj]: This has been the default for a very long time.

File: Functions/Example/FirebaseFunctions.xcodeproj/project.pbxproj (hunk 10 above)
Comment (bstpierr) [reply Functions/Example/FirebaseFunctions.xcodeproj/project.pbxproj]: Re-added in latest commit.

**Comment by bstpierr** [comment]
This should ready to review again. I was able to get everything compiling and passing once I could iterate quickly with `build.sh`.

File: Functions/Example/FirebaseFunctions.xcodeproj/project.pbxproj [hunk 11]
```diff
@@ -28,8 +28,9 @@
 		7CBFAA88205702AB00A65866 /* Foundation.framework in Frameworks */ = {isa = PBXBuildFile; fileRef = 6003F58D195388D20070C39A /* Foundation.framework */; };
 		7CBFAA8B205702AB00A65866 /* InfoPlist.strings in Resources */ = {isa = PBXBuildFile; fileRef = 6003F5B8195388D20070C39A /* InfoPlist.strings */; };
 		7CF3BEC21F97EE2F00B16B6E /* FUNFakeInstanceID.m in Sources */ = {isa = PBXBuildFile; fileRef = 7CF3BEC01F97EE2F00B16B6E /* FUNFakeInstanceID.m */; };
-		7CF563091F1FE70600FEE1F4 /* FUNFakeApp.m in Sources */ = {isa = PBXBuildFile; fileRef = 7CF563081F1FE70600FEE1F4 /* FUNFakeApp.m */; };
 		873B8AEB1B1F5CCA007FD442 /* Main.storyboard in Resources */ = {isa = PBXBuildFile; fileRef = 873B8AEA1B1F5CCA007FD442 /* Main.storyboard */; };
+		C8E1AD6521B5ABC300A43E79 /* FUNFakeApp.m in Sources */ = {isa = PBXBuildFile; fileRef = 7CF563081F1FE70600FEE1F4 /* FUNFakeApp.m */; };
```
[the remark this answers is not in the document]
Comment (bstpierr) [reply Functions/Example/FirebaseFunctions.xcodeproj/project.pbxproj]: I think it's because I moved it from the single app target to the two test targets to avoid having the private imports in the search path of the example app.

File: Functions/FirebaseFunctions/FIRFunctions.m [hunk 12]
```diff
@@ -31,12 +37,19 @@
...
+@end
 
-@interface FIRFunctions () {
+@interface FIRFunctions () <FIRComponentRegistrant, FIRFunctionsInstanceProvider> {
   // The network client to use for http requests.
   GTMSessionFetcherService *_fetcherService;
   // The projectID to use for all function references.
   FIRApp *_app;
+  // The Auth instance to use, may be nil if auth is unavailable.
+  id<FIRAuthInterop> _auth;
```
Comment (ryanwilson) [inline Functions/FirebaseFunctions/FIRFunctions.m]: Nit: It doesn't look like we use this var anywhere other than in the constructor, so we can probably get rid of it!

File: Functions/FirebaseFunctions/FIRFunctions.m [hunk 13]
```diff
@@ -75,16 +101,23 @@ + (instancetype)functionsForApp:(FIRApp *)app region:(NSString *)region {
...
     if (!region) {
       FUNThrowInvalidArgument(@"FIRFunctions region cannot be nil.");
     }
     _fetcherService = [[GTMSessionFetcherService alloc] init];
     _app = app;
+    _auth = auth;
     _region = [region copy];
     _serializer = [[FUNSerializer alloc] init];
-    _contextProvider = [[FUNContextProvider alloc] initWithApp:app];
+    _contextProvider = [[FUNContextProvider alloc] initWithAuth:_auth];
```
Comment (ryanwilson) [inline Functions/FirebaseFunctions/FIRFunctions.m]: Nit: Along with the above, we can just use `auth` instead of assigning it to an instance var since it's not used outside of this scope.

**Comment by bstpierr** [comment]
Hi Bryan, could you have a look at this CL which moves Functions to the new Interop platform.

File: Functions/FirebaseFunctions/FIRFunctions.m (hunk 12 above)
Comment (bstpierr) [reply Functions/FirebaseFunctions/FIRFunctions.m]: Done.

File: Functions/FirebaseFunctions/FIRFunctions.m (hunk 13 above)
Comment (bstpierr) [reply Functions/FirebaseFunctions/FIRFunctions.m]: Done.

File: Functions/Example/IntegrationTests/FIRIntegrationTests.m [hunk 14]
```diff
@@ -77,7 +81,10 @@ - (void)testScalar {
 - (void)testToken {
   // Recreate _functions with a token.
   id app = [[FUNFakeApp alloc] initWithProjectID:@"functions-integration-test" token:@"token"];
-  FIRFunctions *functions = [FIRFunctions functionsForApp:app];
+  FIRFunctions *functions = [[FIRFunctions alloc]
+      initWithApp:app
+           region:@"my-region"
+             auth:[[FIRAuthInteropFake alloc] initWithToken:nil userID:nil error:nil]];
```
Comment (bklimt) [inline Functions/Example/IntegrationTests/FIRIntegrationTests.m]: I don't understand why this passes. Shouldn't functions be using the `FIRAuthInteropFake`'s token instead of the one passed into the `FUNFakeApp`?

File: Functions/Example/Tests/FIRFunctionsTests.m [hunk 15]
```diff
@@ -34,7 +34,7 @@ - (void)tearDown {
 
 - (void)testURLWithName {
   id app = [[FUNFakeApp alloc] initWithProjectID:@"my-project"];
-  FIRFunctions *functions = [FIRFunctions functionsForApp:app region:@"my-region"];
+  FIRFunctions *functions = [[FIRFunctions alloc] initWithApp:app region:@"my-region" auth:nil];
```
Comment (bklimt) [inline Functions/Example/Tests/FIRFunctionsTests.m]: Can this keep the constructor it had? It's nice to have a test of the constructor, and it doesn't seem like we need to pass auth here.

File: Functions/FirebaseFunctions/FIRFunctions.m [hunk 16]
```diff
@@ -75,6 +99,12 @@ + (instancetype)functionsForApp:(FIRApp *)app region:(NSString *)region {
 }
 
 - (instancetype)initWithApp:(FIRApp *)app region:(NSString *)region {
+  return [self initWithApp:app region:region auth:FIR_COMPONENT(FIRAuthInterop, _app.container)];
+}
+
+- (instancetype)initWithApp:(FIRApp *)app
```
Comment (bklimt) [inline Functions/FirebaseFunctions/FIRFunctions.m]: Is this the `NS_DESIGNATED_INITIALIZER`? If so, please mark it as such.

Now that we only use `app` to get the project ID, can we change this to `initWithProjectID:(NSString *)projectID` and keep that in an ivar instead of the app? Then we could get rid of `FUNFakeApp` completely.

File: Functions/Example/IntegrationTests/FIRIntegrationTests.m (hunk 14 above)
Comment (bstpierr) [reply Functions/Example/IntegrationTests/FIRIntegrationTests.m]: I'm also unsure how this was passing.

It looks like [`callFunction`](https://github.com/firebase/firebase-ios-sdk/blob/master/Functions/FirebaseFunctions/FIRFunctions.m#L163) would have an empty Authorization header, but maybe the mock server still returns an empty data response with a 200 OK?

File: Functions/Example/Tests/FIRFunctionsTests.m (hunk 15 above)
Comment (bstpierr) [reply Functions/Example/Tests/FIRFunctionsTests.m]: By removing the app from the designated initializer and deleting `FUNFakeApp`, this class factory method no longer works for this test.

Since the test is for `URLWithName:`, this seems reasonable to me. If we wish to test the behaviour of the convenience initializers, I would recommend separate tests using OCMock to modify the default app or use a partial mock of an app that verifies the given fields are accessed as expected instead of having a custom fake app instance.

File: Functions/FirebaseFunctions/FIRFunctions.m (hunk 16 above)
Comment (bstpierr) [reply Functions/FirebaseFunctions/FIRFunctions.m]: I like this! The fewer references to app instances there are, the better.

File: Functions/Example/IntegrationTests/FIRIntegrationTests.m (hunk 14 above)
Comment (bstpierr) [reply Functions/Example/IntegrationTests/FIRIntegrationTests.m]: Actually, I bet it wasn't passing. Travis doesn't run the integration tests on PRs, just the master branch. I'll run them locally to make sure they're passing now.

File: Functions/Example/Tests/FIRFunctionsTests.m (hunk 15 above)
Comment (bklimt) [reply Functions/Example/Tests/FIRFunctionsTests.m]: Yes, it makes sense for this to change now that `FUNFakeApp` is going away.

File: Functions/Example/IntegrationTests/FIRIntegrationTests.m (hunk 14 above)
Comment (bstpierr) [reply Functions/Example/IntegrationTests/FIRIntegrationTests.m]: I'm having difficulty running the integration tests. I get many warnings when running [`Backend/start.sh` instructions](https://github.com/firebase/firebase-ios-sdk/tree/master/Functions). See this [gist](https://gist.github.com/bstpierr/9c69a40e7c5f55142671f6e9575c1bf5) of the output I get.

Comment (bstpierr) [reply Functions/Example/IntegrationTests/FIRIntegrationTests.m]: I changed my system to node 8, and ran through the suggestions it printed out, and it seems as though it's connecting to the emulator, but I'm not sure it's behaving as expected.

For example, the `testData` test case gets the error: `Function dataTest in location my-region in project functions-integration-test does not exist` in its [`FIRHTTPSCallableResult` callback](https://github.com/firebase/firebase-ios-sdk/blob/2c9f881b674ebf9b516f63c0b7ec78d9dd61fabc/Functions/Example/IntegrationTests/FIRIntegrationTests.m#L56).

From the emulator's output, it looks like `my-region` isn't what's expected, but rather `us-central1` since it outputs `http://localhost:5005/functions-integration-test/us-central1/dataTest` as the resource.

Switching the tests to using `us-central1` as the region has all the tests but one passing. The [`testNull`](https://github.com/firebase/firebase-ios-sdk/blob/2c9f881b674ebf9b516f63c0b7ec78d9dd61fabc/Functions/Example/IntegrationTests/FIRIntegrationTests.m#L106) attempts to wait on an [expectation twice](https://github.com/firebase/firebase-ios-sdk/blob/2c9f881b674ebf9b516f63c0b7ec78d9dd61fabc/Functions/Example/IntegrationTests/FIRIntegrationTests.m#L124), which is an API violation that throws an error.

Are the integration tests expected to pass? I'm not sure how they would on master without the region change and the duplicate `waitForExpectation:` call.

Comment (bstpierr) [reply Functions/Example/IntegrationTests/FIRIntegrationTests.m]: It appears as though the [last green run of master](https://travis-ci.org/firebase/firebase-ios-sdk/jobs/464154830) compiled the integration tests, but only ran the unit tests.

It seems as though the integration tests [aren't being run on Travis](https://github.com/firebase/firebase-ios-sdk/blob/master/scripts/build.sh#L191).

I'll open an internal bug to fix and enable theses.

**Comment by bstpierr** [comment]
Thank you for the comments @bklimt. Please have a look at my [comments](https://github.com/firebase/firebase-ios-sdk/pull/2113#discussion_r239474202) regarding the integration tests. I think I got to the root of what is happening.

File: Functions/Example/IntegrationTests/FIRIntegrationTests.m (hunk 14 above)
Comment (bklimt) [reply Functions/Example/IntegrationTests/FIRIntegrationTests.m]: > From the emulator's output, it looks like my-region isn't what's expected, but rather us-central1

Yes, the shell script in Backend/ that installs the functions says to install them in `us-central1`. Not sure when that broke.
